step_counter/src/index.ts

```diff
@@ -0,0 +1,129 @@
+  'step_counter:StepCounter',
+  'A service for counting steps.'
+);
+
+// This class holds step count data/utilities. An instance of
+// this class will serve as the service object associated with
+// the StepCounter token (Other developers can substitute their
+// own implementation of a StepCounterItem instead of using this
+// one, by becoming a provider of the StepCounter token).
+class Counter implements StepCounterItem {
+
+  _stepCount: number;
+  countChanged: Signal<any, number>;
+
+  constructor() {
+    this._stepCount = 0;
+    this.countChanged = new Signal<this, number>(this);
+  }
+
+  incrementStepCount(count: number) {
+    this._stepCount += count;
+    this.countChanged.emit(this._stepCount);
+  }
+
+  getStepCount() {
+    return this._stepCount;
+  }
+}
+
+// This plugin is a "provider" in JupyterLab's provider-consumer pattern.
+// For a plugin to become a provider, it must list the token it wants to
+// provide a service object for in its "provides" list, and then it has
+// to return that object (in this case, an instance of the example Counter
+// class defined above) from the function supplied as its activate property.
+// It also needs to supply the interface (the one the service object
+// implements) to JupyterFrontEndPlugin when it's defined.
+const plugin: JupyterFrontEndPlugin<StepCounterItem> = {
+  id: 'step_counter:provider_plugin',
+  description: 'Provider plugin for the step_counter\'s "counter" service object.',
+  autoStart: true,
+  provides: StepCounter,
+  // The activate function here will be called by JupyterLab when the plugin loads
+  activate: (app: JupyterFrontEnd) => {
+    console.log('JupyterLab extension (step_counter/provider plugin) is activated!');
+    const counter = new Counter();
+
+    // Since this plugin "provides" the "StepCounter" service, make sure to
+    // return the object you want to use as the "service object" here (when
+    // other plugins request the StepCounter service, it is this object
+    // that will be supplied)
+    return counter;
+  }
+};
+
+// The Counter class here should be marked as a singleton in this project's
+// package.json according to the Jupyter extension documentation, both in
+// this provider extension and in the consumer extensions. Under the
+// "Jupyterlab" key you should have a "singleton" key set to true.
+// 
+// "jupyterlab": {
+//   "extension": true,
+//   "outputDir": "step_counter/labextension",
+//   "sharedPackages": {
+//       "step_counter": {
+//           "bundled": false,
+//           "singleton": true
+//       }
+//   }
+// },
+//
+// Read more about that here:
+// https://jupyterlab.readthedocs.io/en/latest/extension/extension_dev.html#providing-a-service
+// https://jupyterlab.readthedocs.io/en/latest/extension/extension_dev.html#requiring-a-service
+// https://jupyterlab.readthedocs.io/en/latest/extension/extension_dev.html#optionally-using-a-service
+
+export { StepCounter, StepCounterItem };
```

Comment:
Oh, you export both here. I guess I do not understand why you used `any` after all. Still let's consider using (and explaining declaration merging).